File: TP/mago/set/set.cpp
#include "set.h"
#include "../common/Common.h"

struct Node {
   Hechizo elem; // el elemento que este nodo almacena
   Node* next; // siguiente nodo de la cadena de punteros
};

struct SetSt {
   int size; // cantidad de elementos del conjunto
   Node* first; // puntero al primer elemento
};

/**
  Invariantes de representacion:
    - size es la cantidad de nodos
    - no hay nodos con hechizos repetidos
**/

/// Proposito: retorna un conjunto de hechizos vacio
/// Costo: O(1)
Set emptyS() {
   Set s = new SetSt;
   s->size = 0;
   s->first = NULL;
   return s;
}

/// Proposito: retorna la cantidad de hechizos
/// Costo: O(1)
int sizeS(Set s) {
   return s->size;
}

/// Proposito: indica si el hechizo pertenece al conjunto
/// Costo: O(h), h = cantidad de hechizos
bool belongsS(Hechizo h, Set s) {
    Node* n = s->first;
    while(n != NULL && !mismoHechizo(h,n->elem)){
        n = n->next;
    }
    return n != NULL;
}

/// Proposito: agrega un hechizo al conjunto
/// Costo: O(h), h = cantidad de hechizos
void addS(Hechizo h, Set s) {
    if(!belongsS(h,s)){
        Node* n = new Node;
        n->elem = h;
        n->next = s->first;
        s->first = n;
        s->size++;
    }
}

/// Proposito: borra un hechizo del conjunto (si no existe no hace nada)
/// Costo: O(h), h = cantidad de hechizos
void removeS(Hechizo h, Set s) {
    Node* anterior = NULL;
    Node* actual = s->first;
   while(actual != NULL && actual->elem != h){
        anterior = actual;
        actual = actual->next;
   }

   if(anterior == NULL){
        delete s->first;
        s->first = NULL;
        s->size = 0;
   }
   else{
        actual = actual->next;
        anterior->next = actual;
        s->size--;
   }
}

/// Proposito: borra toda la memoria consumida por el conjunto (pero no la de los hechizos)
/// Costo: O(n)
void destroyS(Set s) {
    Node* n = s->first;
    while(n != NULL) {
        Node* tmp = n;
        n = n->next;
        delete tmp;
	}
	s = NULL;
}

/// Proposito: retorna un nuevo conjunto que es la union entre ambos (no modifica estos conjuntos)
/// Costo: O(h^2), h = cantidad de hechizos
Set unionS(Set s1, Set s2) {
    Set s = new SetSt;
    s->first = NULL;
    s->size = 0;

    Node* n1 = s1->first;
    Node* n2 = s2->first;

    while(n1 != NULL) {
        Node* n = new Node;
        n->elem = n1->elem;
        n->next = s->first;
        s->size++;
        s->first = n;
        n1 = n1->next;
	}

	while(n2 != NULL){
        if(!belongsS(n2->elem,s1)){
            Node* n = new Node;
            n->elem = n2->elem;
            n->next = s->first;
            s->size++;
            s->first = n;
        }
        n2 = n2->next;
	}
	return s;
}

File: PracticaHaskell/Practica4/Set/Set.hs
module Set(Set, emptyS, addS, belongs, sizeS, removeS, unionS, setToList) where

data Set a = MKS [a] Int deriving (Show) 
-- Set Sin Repetidos
-- el TIPO DE REPRESENTACIÓN 
  -- INV.REP.: NINGUNO! Toda lista es aceptable.	

--O(1)
-- Crea un Set vacío.
emptyS :: Set a 
emptyS = MKS [] 0

--O(n*2)
--Dados un elemento y un Set, agrega el elemento al Set.
addS :: Eq a => a -> Set a -> Set a
addS x (MKS xs n) = if elem x xs
                    then MKS xs n
					else MKS (x:xs) (n+1)

--O(n)
--Dados un elemento y un Set indica si el elemento pertenece al Set.
belongs :: Eq a => a -> Set a -> Bool
belongs x (MKS xs n) = elem x xs

--O(1)
--Devuelve la cantidad de elementos distintos de un Set
sizeS :: Eq a => Set a -> Int
sizeS (MKS xs n)  = n

--O(n)
--Devuelve el Set sin ese elemento
removeS :: Eq a => a -> Set a -> Set a
removeS e (MKS xs n) = MKS (borrar e xs) (n+1)

borrar :: Eq a => a -> [a] -> [a]
borrar n []		= []
borrar n (x:xs) = if n == x
					then borrar n xs
					else x : borrar n xs

--O(n*2)
--Dados dos Sets devuelve un Set con todos los elementos de ambos Sets.
unionS :: Eq a => Set a -> Set a -> Set a
unionS (MKS xs n) (MKS ys m) = MKS (listaSinRepetidos (xs ++ ys)) (length(listaSinRepetidos xs))

listaSinRepetidos :: Eq a => [a] -> [a]
listaSinRepetidos [] = []
listaSinRepetidos (x:xs) = if elem x xs
				then listaSinRepetidos xs
				else x:(listaSinRepetidos xs)

--O(n)
--Dados dos conjuntos devuelve un conjunto con todos los elementos en común entre ambos.
intersectionS :: Eq a => Set a -> Set a -> Set a
intersectionS s (MKS xs n) = intersectionL xs s 

intersectionL :: Eq a => [a] -> Set a -> Set a 
intersectionL [] s 	= emptyS
intersectionL (x:xs) s = if belongs x s 
							then addS x (intersectionL xs s)
							else intersectionL xs s

--O(1)
--Dado un Set devuelve una lista con todos los elementos distintos del Set.
setToList :: Eq a => Set a -> [a]
setToList (MKS xs _) = xs